src/binary/non_blocking/raw_binary_reader.rs

```diff
@@ -0,0 +1,1824 @@
+/// Finally, they can retry the read operation that had previously failed.
+///
+/// Note that if the buffer runs out of data between top level values, this will be interpreted
+/// as the end of the stream. Applications can still add more data to the buffer and resume reading.
+#[derive(Debug)]
+pub struct RawBinaryBufferReader<A: AsRef<[u8]>> {
+    ion_version: (u8, u8),
+    state: ReaderState,
+    buffer: BinaryBuffer<A>,
+    parents: Vec<Container>,
+}
+
+impl<A: AsRef<[u8]>> RawBinaryBufferReader<A> {
+    /// Constructs a RawBinaryBufferReader from a value that can be viewed as a byte slice.
+    pub fn new(source: A) -> RawBinaryBufferReader<A> {
+        RawBinaryBufferReader {
+            ion_version: (1, 0),
+            state: ReaderState::Ready,
+            buffer: BinaryBuffer::new(source),
+            parents: Vec::new(), // Does not allocate yet
+        }
+    }
+
+    /// Constructs a disposable view of the buffer's current contents that can be used to find the
+    /// next item in the stream. If the TxReader encounters a problem like invalid or incomplete
+    /// data, it can be discarded without affecting the RawBinaryBufferReader that created it.
+    fn transaction_reader(&mut self) -> TxReader<A> {
+        // Temporarily break apart `self` to get simultaneous mutable references to the buffer,
+        // the reader state, and the parents.
+        let RawBinaryBufferReader {
+            ion_version,
+            state,
+            buffer,
+            parents,
+        } = self;
+
+        // Create a slice of the main buffer that has its own records of how many bytes have been
+        // read and how many remain.
+        let tx_buffer = buffer.slice();
+        TxReader {
+            ion_version: *ion_version,
+            state,
+            main_buffer: buffer,
+            parent: parents.last(),
+            tx_buffer,
+            encoded_value: Default::default(),
+            nop_bytes_count: 0,
+        }
+    }
+
+    /// Moves the reader to the first byte of the next item in the stream.
+    /// If it succeeds, the reader's state will be [ReaderState::Ready].
+    /// If there is not enough data in the buffer to reach the next item, the reader's state will
+    /// be [ReaderState::Skipping], indicating that it is mid-stream and awaiting more data.
+    fn advance_to_next_item(&mut self) -> IonResult<()> {
+        use ReaderState::*;
+        match &mut self.state {
+            // We're done advancing and are ready to try reading the current byte.
+            Ready => Ok(()),
+            // The `Skipping(n)` state means that we ran out of data on a previous call to `next()`.
+            // The only way to fix it is to add more input bytes to the buffer using the `append_bytes`
+            // or `read_from` methods. Adding input bytes to the buffer is only supported
+            // when the generic type `A` is `Vec<u8>`.
```

Comment:
🗺️ There's a tradeoff lurking here. Moving to the `Skipping(n)` state means that we leave the reader in no man's land; no operation will succeed until more data has been added.

We could instead have the reader simply return `IonError::Incomplete` without changing its state. However:
* This would mean that the only time the reader could advance is when the entire current value is in the buffer. Given that some real-world use cases work with multi-GB values, this feels like an onerous requirement. Having  a `Skipping(n)` state allows the reader to incrementally skip over very large values.
* If a call to `next()` (which in turn calls `advance_to_next_item()`) fails as `Incomplete`, I don't think users would expect to be able to perform more read operations on the value they were on before calling `next()`.

If a use case for the alternative behavior surfaces down the road, I could see adding a config option for this (`disable_incremental_advancement`?).